controllers/volumereplication_controller.go

```diff
@@ -209,26 +211,41 @@ func (r *VolumeReplicationReconciler) SetupWithManager(mgr ctrl.Manager, cfg *co
 }
 
 // markVolumeAsPrimary defines and runs a set of tasks required to mark a volume as primary
-func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeID string, parameters, secrets map[string]string) (string, error) {
+func (r *VolumeReplicationReconciler) markVolumeAsPrimary(volumeID string, forcePromote bool, parameters, secrets map[string]string) (string, error) {
 	c := replication.CommonRequestParameters{
 		VolumeID:    volumeID,
 		Parameters:  parameters,
 		Secrets:     secrets,
 		Replication: r.Replication,
 	}
-	force := false
 	var markVolumeAsPrimaryTasks = []*tasks.TaskSpec{
 		{
 			Name: enableVolumeReplication,
 			Task: replication.NewEnableTask(c),
 		},
 		{
 			Name: promoteVolume,
-			Task: replication.NewPromoteVolumeTask(c, force),
+			Task: replication.NewPromoteVolumeTask(c, forcePromote),
+			KnownErrors: []codes.Code{
+				codes.FailedPrecondition,
+			},
 		},
 	}
 
-	return tasks.RunAll(markVolumeAsPrimaryTasks)
+	taskName, taskError := tasks.RunAll(markVolumeAsPrimaryTasks)
+	if taskError.Position >= 0 {
+		s, ok := status.FromError(taskError.Message)
+		if !ok {
+			return taskName, taskError.Message
+		}
+		for _, e := range markVolumeAsPrimaryTasks[taskError.Position].KnownErrors {
+			if s.Code() == e {
+				r.markVolumeAsPrimary(volumeID, true, parameters, secrets)
```

Comment:
This will call the EnableReplication again. we can avoid one more network call if we take this approach https://github.com/Madhu-1/operator/commit/6ab7b66c67b78c3ceffd31b4fb8c70e01665bfac

Reply:
Oh I forgot that we're dealing with GRPC calls. I'll fix it.